win: add prompt to tools installation script

joaocgreis opened the pull request:
Since the script to install tools for native modules (https://github.com/nodejs/node/pull/22645) was included in releases, we've got several comments from users surprised by reboots.

This adds a more visible warning and a prompt to lessen the probability of users skipping ahead without reading.

Fixes: https://github.com/nodejs/Release/issues/369

cc @nodejs/platform-windows 

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

joaocgreis commented:
CI: https://ci.nodejs.org/view/All/job/node-test-pull-request/18252/

jdalton commented on tools/msvs/install_tools/install_tools.bat:
```diff
@@ -20,8 +22,15 @@ echo.
+echo Please type YES followed by enter to confirm that you have saved all your
+set /p "ACCEPT_PROMPT=work and closed all open programs: "
+if /i not "%ACCEPT_PROMPT%"=="yes" (
+  echo Please type YES or close the window to exit.
```
`Please type YES or close the window to exit.` reads like typing Yes or closing the window will exit. I think the `type Yes` bit should read more like the initial prompt to `type YES to confirm or...`.

joaocgreis commented:
@jdalton thanks! Updated.

ferventcoder commented:
I like the type yes to confirm, but this does not mention the UAC bit - I would like to see the wording from #24000 with this yes prompt

ferventcoder commented:
Also I missed this was here on initial search and filed my own PR - apologies on that. We would like the wording and notes from the other PR brought into here to address that UAC will be disabled and make the user aware of what they will need to do should they stop the process instead of letting it finish

joaocgreis commented:
@ferventcoder your PR adds a lot of good info, thanks for it! I included it here, let me know if this was not what you had in mind. I took the opportunity to change the commit message (subsystem and 72-col lines) and added a fixup addressing feedback from @Trott. I also moved my prompt to the new screen you created. Let me know if there's any change you'd like to make.

ferventcoder commented:
@joaocgreis looks great, thank you!

ferventcoder commented on tools/msvs/install_tools/install_tools.bat:
```diff
@@ -47,7 +49,31 @@ echo script is at your own risk. Please read the Chocolatey's legal terms of use
+echo in fact it is just a normal Windows Updates reboot cycle.
+echo.
+echo If this is not what you would like to occur, you can close this window
+echo to stop now.
```
Perhaps we can pull these two lines out since you reference this below.

ferventcoder commented on tools/msvs/install_tools/install_tools.bat:
```diff
@@ -47,7 +49,31 @@ echo script is at your own risk. Please read the Chocolatey's legal terms of use
+echo Your computer may REBOOT SEVERAL TIMES WITHOUT FURTHER WARNING.
+echo Please type YES followed by enter to confirm that you have saved all your
+set /p "ACCEPT_PROMPT=work and closed all open programs: "
+if /i not "%ACCEPT_PROMPT%"=="yes" (
```
Does this handle case sensitivity? You are asking for "YES", and this accepts "yes"

joaocgreis replied:
The `/i` switch is for ignoring case. I used uppercase above to make it obvious that only `yes` needs to be typed, but the user can type in any case.

joaocgreis commented:
@ferventcoder thanks, updated!

@jdalton this changed a lot since you reviewed. I'll assume your approval still holds, please let me know if not.

joaocgreis commented:
If there are no objections, I will land this tomorrow.

Full CI: https://ci.nodejs.org/view/All/job/node-test-pull-request/18291/

Trott commented:
Landed in 315b1c6...ab2778d

MylesBorins commented:
@trott would you be willing to backport to 10.x

Lmk if it lands cleanly

MylesBorins commented:
@Trott I've landed this in 10.x-staging and it should make it out in the next 10.x release

joaocgreis commented:
@Trott @MylesBorins thanks!

MPagel commented:
The new prompt/dialogue is helpful, but it is insufficient.

On my system I have one username for everyday access and install programs using a different user account using UAC elevation while still in the GUI for the low-access user.  This is SOP for employees of my university system (with approximately 200 000 users).

After reboot, choco required me to be logged in to the GUI as the Admin user (a no-no for my employer), without any message on my standard user screen, before releasing the UAC back to normal.  As I and my almost quarter million co-workers could technically be fired for knowingly violating security SOPs, I'd strongly urge you to consider not using choco until such time as a) post-reboot install steps take place even without a logged in user b) reboots can be suppressed until such time as the user sees fit and c) UAC is never disabled/circumvented.

Assuming that you are sticking with choco as your standard install method, at the absolute very least, there should be a line in the prompt/dialogue that specifies that a log in to the account that is installing the program will need to be performed after the reboot.  If possible, the install should detect the GUI username as well as the installing username, and if not a match, popping up an alert or message window on the GUI desktop after the reboot but before the install process is complete.

refack commented:
I think we should pull this out of v10. It's not LTS ready.

MylesBorins commented:
@refack can you open a revert against v10.x-staging
